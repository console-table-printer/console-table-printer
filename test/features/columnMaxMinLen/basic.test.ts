import { Table } from '../../../index';
import { getTableBody, getTableHeader } from '../../testUtils/getRawData';

describe('Column Max/Min Length Tests', () => {
  it('should handle basic column maxLen', () => {
    const p = new Table()
      .addColumn({ name: 'limited', maxLen: 10 })
      .addColumn({ name: 'unlimited' });

    p.addRows([
      { limited: 'This text is longer than 10 chars', unlimited: 'Short' },
      { limited: 'Short', unlimited: 'This can be as long as it needs to be' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle basic column minLen', () => {
    const p = new Table()
      .addColumn({ name: 'padded', minLen: 15 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { padded: 'Hi', normal: 'Normal text' },
      { padded: 'Hello', normal: 'Another row' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle both maxLen and minLen together', () => {
    const p = new Table()
      .addColumn({ name: 'constrained', minLen: 10, maxLen: 20 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { constrained: 'Short', normal: 'Normal column' },
      { constrained: 'This is exactly twenty chars plus more', normal: 'Another' },
      { constrained: 'Medium length', normal: 'Row 3' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle maxLen with different alignments', () => {
    const p = new Table()
      .addColumn({ name: 'left', maxLen: 10, alignment: 'left' })
      .addColumn({ name: 'center', maxLen: 10, alignment: 'center' })
      .addColumn({ name: 'right', maxLen: 10, alignment: 'right' });

    p.addRows([
      {
        left: 'This will be wrapped to multiple lines',
        center: 'Center aligned text wrapping',
        right: 'Right aligned text wrapping'
      },
      { left: 'Short', center: 'Mid', right: 'End' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle minLen with different alignments', () => {
    const p = new Table()
      .addColumn({ name: 'left', minLen: 15, alignment: 'left' })
      .addColumn({ name: 'center', minLen: 15, alignment: 'center' })
      .addColumn({ name: 'right', minLen: 15, alignment: 'right' });

    p.addRows([
      { left: 'L', center: 'C', right: 'R' },
      { left: 'Left', center: 'Center', right: 'Right' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle edge case where maxLen equals text length', () => {
    const p = new Table()
      .addColumn({ name: 'exact', maxLen: 5 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { exact: '12345', normal: 'Exactly 5 chars' },
      { exact: '1234', normal: 'Less than 5' },
      { exact: '123456', normal: 'More than 5' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle edge case where minLen equals text length', () => {
    const p = new Table()
      .addColumn({ name: 'exact', minLen: 5 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { exact: '12345', normal: 'Exactly 5 chars' },
      { exact: '1234', normal: 'Less than 5' },
      { exact: '123456', normal: 'More than 5' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle very small maxLen values', () => {
    const p = new Table()
      .addColumn({ name: 'tiny', maxLen: 1 })
      .addColumn({ name: 'small', maxLen: 3 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { tiny: 'Word', small: 'Testing', normal: 'Normal column' },
      { tiny: 'A', small: 'OK', normal: 'Second row' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle numeric values with maxLen and minLen', () => {
    const p = new Table()
      .addColumn({ name: 'numbers', minLen: 10, maxLen: 15 });

    p.addRows([
      { numbers: 42 },
      { numbers: 3.14159265359 },
      { numbers: 1000000000000000 },
      { numbers: 0 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle empty values with minLen', () => {
    const p = new Table()
      .addColumn({ name: 'padded', minLen: 10 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { padded: '', normal: 'Empty padded' },
      { padded: null, normal: 'Null padded' },
      { padded: undefined, normal: 'Undefined padded' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle maxLen/minLen with colored columns', () => {
    const p = new Table()
      .addColumn({ name: 'colored', color: 'red', maxLen: 10, minLen: 5 })
      .addColumn({ name: 'normal' });

    p.addRows([
      { colored: 'Hi', normal: 'Short colored text' },
      { colored: 'This is a very long text that should wrap', normal: 'Long colored text' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle global maxLen and minLen from table options', () => {
    const p = new Table({
      columns: [
        { name: 'col1' },
        { name: 'col2' },
        { name: 'col3', maxLen: 5 }, // Override global maxLen
      ],
      defaultColumnOptions: {
        maxLen: 12,
        minLen: 8,
      }
    });

    p.addRows([
      { col1: 'Short', col2: 'Medium text', col3: 'Override' },
      { col1: 'This is longer than 12 characters', col2: 'Hi', col3: 'Test' },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });
});