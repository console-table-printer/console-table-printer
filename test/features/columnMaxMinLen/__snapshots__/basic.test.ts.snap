// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Column Max/Min Length Tests should handle basic column maxLen 1`] = `
"┌────────────┬───────────────────────────────────────┐
│[37m [0m[01m   limited[0m │[37m [0m[01m                            unlimited[0m │
├────────────┼───────────────────────────────────────┤
│[37m [0m[37m This text[0m │[37m [0m[37m                                Short[0m │
│[37m [0m[37m is longer[0m │[37m [0m[37m                                     [0m │
│[37m [0m[37m   than 10[0m │[37m [0m[37m                                     [0m │
│[37m [0m[37m     chars[0m │[37m [0m[37m                                     [0m │
│[37m [0m[37m     Short[0m │[37m [0m[37mThis can be as long as it needs to be[0m │
└────────────┴───────────────────────────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle basic column minLen 1`] = `
"┌─────────────────┬─────────────┐
│[37m [0m[01m         padded[0m │[37m [0m[01m     normal[0m │
├─────────────────┼─────────────┤
│[37m [0m[37m             Hi[0m │[37m [0m[37mNormal text[0m │
│[37m [0m[37m          Hello[0m │[37m [0m[37mAnother row[0m │
└─────────────────┴─────────────┘"
`;

exports[`Column Max/Min Length Tests should handle both maxLen and minLen together 1`] = `
"┌──────────────────────┬───────────────┐
│[37m [0m[01m         constrained[0m │[37m [0m[01m       normal[0m │
├──────────────────────┼───────────────┤
│[37m [0m[37m               Short[0m │[37m [0m[37mNormal column[0m │
│[37m [0m[37m     This is exactly[0m │[37m [0m[37m      Another[0m │
│[37m [0m[37m   twenty chars plus[0m │[37m [0m[37m             [0m │
│[37m [0m[37m                more[0m │[37m [0m[37m             [0m │
│[37m [0m[37m       Medium length[0m │[37m [0m[37m        Row 3[0m │
└──────────────────────┴───────────────┘"
`;

exports[`Column Max/Min Length Tests should handle edge case where maxLen equals text length 1`] = `
"┌────────┬─────────────────┐
│[37m [0m[01m exact[0m │[37m [0m[01m         normal[0m │
├────────┼─────────────────┤
│[37m [0m[37m 12345[0m │[37m [0m[37mExactly 5 chars[0m │
│[37m [0m[37m  1234[0m │[37m [0m[37m    Less than 5[0m │
│[37m [0m[37m123456[0m │[37m [0m[37m    More than 5[0m │
└────────┴─────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle edge case where minLen equals text length 1`] = `
"┌────────┬─────────────────┐
│[37m [0m[01m exact[0m │[37m [0m[01m         normal[0m │
├────────┼─────────────────┤
│[37m [0m[37m 12345[0m │[37m [0m[37mExactly 5 chars[0m │
│[37m [0m[37m  1234[0m │[37m [0m[37m    Less than 5[0m │
│[37m [0m[37m123456[0m │[37m [0m[37m    More than 5[0m │
└────────┴─────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle empty values with minLen 1`] = `
"┌────────────┬──────────────────┐
│[37m [0m[01m    padded[0m │[37m [0m[01m          normal[0m │
├────────────┼──────────────────┤
│[37m [0m[37m          [0m │[37m [0m[37m    Empty padded[0m │
│[37m [0m[37m          [0m │[37m [0m[37m     Null padded[0m │
│[37m [0m[37m          [0m │[37m [0m[37mUndefined padded[0m │
└────────────┴──────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle global maxLen and minLen from table options 1`] = `
"┌──────────────┬──────────────┬──────────┐
│[37m [0m[01m        col1[0m │[37m [0m[01m        col2[0m │[37m [0m[01m    col3[0m │
├──────────────┼──────────────┼──────────┤
│[37m [0m[37m       Short[0m │[37m [0m[37m Medium text[0m │[37m [0m[37mOverride[0m │
│[37m [0m[37m     This is[0m │[37m [0m[37m          Hi[0m │[37m [0m[37m    Test[0m │
│[37m [0m[37m longer than[0m │[37m [0m[37m            [0m │[37m [0m[37m        [0m │
│[37m [0m[37m          12[0m │[37m [0m[37m            [0m │[37m [0m[37m        [0m │
│[37m [0m[37m  characters[0m │[37m [0m[37m            [0m │[37m [0m[37m        [0m │
└──────────────┴──────────────┴──────────┘"
`;

exports[`Column Max/Min Length Tests should handle maxLen with different alignments 1`] = `
"┌────────────┬────────────┬────────────┐
│[37m [0m[01mleft      [0m │[37m [0m[01m  center  [0m │[37m [0m[01m     right[0m │
├────────────┼────────────┼────────────┤
│[37m [0m[37mThis will [0m │[37m [0m[37m  Center  [0m │[37m [0m[37m     Right[0m │
│[37m [0m[37mbe wrapped[0m │[37m [0m[37m aligned  [0m │[37m [0m[37m   aligned[0m │
│[37m [0m[37mto        [0m │[37m [0m[37m   text   [0m │[37m [0m[37m      text[0m │
│[37m [0m[37mmultiple  [0m │[37m [0m[37m wrapping [0m │[37m [0m[37m  wrapping[0m │
│[37m [0m[37mlines     [0m │[37m [0m[37m          [0m │[37m [0m[37m          [0m │
│[37m [0m[37mShort     [0m │[37m [0m[37m   Mid    [0m │[37m [0m[37m       End[0m │
└────────────┴────────────┴────────────┘"
`;

exports[`Column Max/Min Length Tests should handle maxLen/minLen with colored columns 1`] = `
"┌────────────┬────────────────────┐
│[37m [0m[01m   colored[0m │[37m [0m[01m            normal[0m │
├────────────┼────────────────────┤
│[37m [0m[31m        Hi[0m │[37m [0m[37mShort colored text[0m │
│[37m [0m[31m This is a[0m │[37m [0m[37m Long colored text[0m │
│[37m [0m[31m very long[0m │[37m [0m[37m                  [0m │
│[37m [0m[31m text that[0m │[37m [0m[37m                  [0m │
│[37m [0m[31m    should[0m │[37m [0m[37m                  [0m │
│[37m [0m[31m      wrap[0m │[37m [0m[37m                  [0m │
└────────────┴────────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle minLen with different alignments 1`] = `
"┌─────────────────┬─────────────────┬─────────────────┐
│[37m [0m[01mleft           [0m │[37m [0m[01m    center     [0m │[37m [0m[01m          right[0m │
├─────────────────┼─────────────────┼─────────────────┤
│[37m [0m[37mL              [0m │[37m [0m[37m       C       [0m │[37m [0m[37m              R[0m │
│[37m [0m[37mLeft           [0m │[37m [0m[37m    Center     [0m │[37m [0m[37m          Right[0m │
└─────────────────┴─────────────────┴─────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle numeric values with maxLen and minLen 1`] = `
"┌──────────────────┐
│[37m [0m[01m         numbers[0m │
├──────────────────┤
│[37m [0m[37m              42[0m │
│[37m [0m[37m   3.14159265359[0m │
│[37m [0m[37m1000000000000000[0m │
│[37m [0m[37m               0[0m │
└──────────────────┘"
`;

exports[`Column Max/Min Length Tests should handle very small maxLen values 1`] = `
"┌──────┬─────────┬───────────────┐
│[37m [0m[01mtiny[0m │[37m [0m[01m  small[0m │[37m [0m[01m       normal[0m │
├──────┼─────────┼───────────────┤
│[37m [0m[37mWord[0m │[37m [0m[37mTesting[0m │[37m [0m[37mNormal column[0m │
│[37m [0m[37m   A[0m │[37m [0m[37m     OK[0m │[37m [0m[37m   Second row[0m │
└──────┴─────────┴───────────────┘"
`;
